### PR TITLE
feat(canvas): add API client, data models, and connector scaffold 1/4

### PR DESCRIPTION
**Pull request by benwu408**
Add CanvasApiClient, data models (CanvasCourse, CanvasPage, CanvasAssignment, CanvasAnnouncement), checkpoint class, and register the Canvas connector source.

## Description

Add the Canvas LMS API client (CanvasApiClient) with HTTPS-only URL validation, safe Link-header pagination with host/scheme verification to prevent bearer token leakage, and comprehensive error mapping (401/403/404/429/5xx → OnyxErrorCode). 

Define Pydantic data models (CanvasCourse, CanvasPage, CanvasAssignment, CanvasAnnouncement) with from_api factory methods, and the CanvasConnectorCheckpoint for staged indexing. 

Includes unit tests for all client methods, error codepaths, and pagination security checks.

## How Has This Been Tested?

Unit tests covering all client methods: URL validation (success + rejection), URL building, header construction, get() success/error paths (all 3 Canvas error shapes, invalid JSON, pagination param skipping), and _parse_next_link (found, not found, empty, multiple rels, host mismatch rejection, non-HTTPS rejection).

## PR Stack

**Stacked PRs on my fork (each targets the previous branch for isolated review):**
1. [#1 — Scaffold: API client, data models, connector registration](https://github.com/benwu408/onyx/pull/1)
2. [#2 — Connector data fetching and conversion](https://github.com/benwu408/onyx/pull/2)
3. [#3 — Checkpoint-based indexing](https://github.com/benwu408/onyx/pull/3)
4. [#4 — Permission sync and frontend registration](https://github.com/benwu408/onyx/pull/4)

**PRs on upstream onyx (all target main):**
1. [#9385 — Scaffold](https://github.com/onyx-dot-app/onyx/pull/9385)
2. [#9386 — Connector data](https://github.com/onyx-dot-app/onyx/pull/9386)
3. [#9387 — Checkpoint](https://github.com/onyx-dot-app/onyx/pull/9387)
4. [#9388 — Perm sync + frontend](https://github.com/onyx-dot-app/onyx/pull/9388)

## Additional Options

- [ ] [Optional] Please cherry-pick this PR to the latest release version.
- [x] [Optional] Override Linear Check

---
## Summary by cubic
Scaffolds the Canvas LMS connector with a secure API client, core models, and a resumable checkpoint. Adds thorough client tests and tightens pagination, URL validation, and error handling.

- **New Features**
  - `CanvasApiClient` for authenticated GETs under `/api/v1` with a 30s timeout, https-only base URL validation/normalization, Link-header pagination (validated host + https), a URL builder, optional `endpoint` when using `full_url`, and HTTP status-to-`OnyxErrorCode` mapping with detailed error extraction via `OnyxError`. Clarifies in docs that `_build_url` is only used on first requests.
  - Models and state: `CanvasCourse`, `CanvasPage` (uses `page_id`), `CanvasAssignment`, `CanvasAnnouncement`; `CanvasConnectorCheckpoint` with `course_ids`, `current_course_index`, `stage` (`pages` | `assignments` | `announcements`), `next_url`, and `advance_course()`.

- **Bug Fixes**
  - Pagination/URL safety: validate host and https on `next` links, cache expected host, and skip `params` when using `full_url`.
  - Error handling: raise on invalid JSON for 2xx, log on non-2xx JSON failures, and map unrecognized 4xx to `BAD_GATEWAY`.
  - Tests: fix mock response helper to treat empty dicts correctly.

<sup>Written for commit 3c2476ec6fc0f77a2e8fea5b31fa0f32f82c0093. Summary will update on new commits.</sup>

### CODE REVIEW

File: backend/onyx/connectors/canvas/access.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+from onyx.access.models import ExternalAccess
+from onyx.connectors.canvas.client import CanvasApiClient
+
+
+def get_course_permissions(
+    canvas_client: CanvasApiClient,
+    course_id: int,
+) -> ExternalAccess | None:
+    return None
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/access.py]: This is extremely important

File: backend/onyx/connectors/canvas/connector.py
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:1]: The imports of this file need to be cleaned up

File: backend/onyx/connectors/canvas/connector.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class CanvasPage(BaseModel):
+    page_id: int
+    url: str
+    title: str
+    body: str | None
+    created_at: str
+    updated_at: str
+    course_id: int
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/connector.py:24]: I believe this may be incorrect but if there are specific docs you are referencing please share @greptile.

[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/connector.py:24]: @benwu408 if this is indeed true that you don't need the html_url you should ensure that this design is aligned on all of the classes that you have implemented

File: backend/onyx/connectors/canvas/connector.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    Fields:
+        course_ids: Materialized list of course IDs to process.
+        current_course_index: Index into course_ids for current course.
+        stage: Which item type we're processing for the current course.
+    """
+
+    course_ids: list[int] = []
+    current_course_index: int = 0
+    stage: CanvasStage = "pages"
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/connector.py:67]: I agree with this comment, we need to be fault tolerant to any failures that may occur so we should try to figure out where exactly we failed and ensure that we can pick up from that point forward

File: backend/onyx/connectors/canvas/client.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+_CANVAS_CALL_TIMEOUT = 30
+_CANVAS_API_VERSION = "/api/v1"
+
+
+class CanvasClientRequestFailedError(ConnectionError):
+    def __init__(self, message: str, status_code: int):
+        super().__init__(
+            f"Canvas API request failed with status {status_code}: {message}"
+        )
+        self.status_code = status_code
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py]: We now have https://github.com/onyx-dot-app/onyx/blob/11cfc92f159458571bd02f35417d2f24b238d27e/backend/onyx/error_handling/error_codes.py#L14 so in theory we should try to reference the code from there. Might be harder to refactor this function but longer term it will be more aligned with our codebase

File: backend/onyx/connectors/canvas/client.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class CanvasApiClient:
+    def __init__(
+        self,
+        bearer_token: str,
+        canvas_base_url: str,
+    ) -> None:
+        self.bearer_token = bearer_token
+        self.base_url = canvas_base_url.rstrip("/") + _CANVAS_API_VERSION
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py]: Do we need to do any validation or is this validation helping elsewhere?

File: backend/onyx/connectors/canvas/client.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    def get(
+        self,
+        endpoint: str,
+        params: dict[str, Any] | None = None,
+        full_url: str | None = None,
+    ) -> tuple[Any, str | None]:
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/client.py:71]: This is a valid comment

File: backend/onyx/connectors/canvas/client.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        """Make a GET request to the Canvas API.
+
+        Returns a tuple of (json_body, next_url).
+        next_url is parsed from the Link header and is None if there are no more pages.
+        If full_url is provided, it is used directly (for following pagination links).
+        """
+        url = full_url if full_url else self._build_url(endpoint)
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/client.py:85]: Goes back to the comment I left above about validating proper inputs

File: backend/onyx/connectors/canvas/client.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+import re
+from typing import Any
+
+from onyx.connectors.cross_connector_utils.rate_limit_wrapper import (
+    rl_requests,
+)
+
+_CANVAS_CALL_TIMEOUT = 30
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py]: What is this variable? In seconds, minutes, hours?

File: backend/onyx/connectors/canvas/client.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        headers = self._build_headers()
+
+        response = rl_requests.get(
+            url, headers=headers, params=params, timeout=_CANVAS_CALL_TIMEOUT
+        )
+
+        try:
+            json = response.json()
+        except Exception:
+            json = {}
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/client.py]: Should be properly addressed

File: backend/onyx/connectors/canvas/client.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        url = full_url if full_url else self._build_url(endpoint)
+        headers = self._build_headers()
+
+        response = rl_requests.get(
+            url, headers=headers, params=params, timeout=_CANVAS_CALL_TIMEOUT
+        )
+
+        try:
+            json = response.json()
+        except Exception:
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/client.py]: We should definitely be logging errors here

File: backend/onyx/connectors/canvas/client.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            json = {}
+
+        if response.status_code >= 300:
+            error = response.reason
+            error_field = json.get("error")
+            if isinstance(error_field, dict):
+                response_error = error_field.get("message", "")
+                if response_error:
+                    error = response_error
+            elif isinstance(error_field, str):
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py]: This means that redirection requests would also get caught by the checks you are doing here. Is that correct? Are we only looking for successful error codes? 

Another place to use the error_codes.py file

File: backend/onyx/connectors/canvas/client.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+        next_url = self._parse_next_link(response.headers.get("Link", ""))
+        return json, next_url
+
+    @staticmethod
+    def _parse_next_link(link_header: str) -> str | None:
+        """Extract the 'next' URL from a Canvas Link header."""
+        for match in re.finditer(r'<([^>]+)>;\s*rel="next"', link_header):
+            return match.group(1)
+        return None
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py]: A regex pattern like this should be added to the top so it has higher visibility. This is hard to read and resolve if it's deep within the file

**Review by justin-tahara — COMMENTED** [review]
Leaving initial comments

File: backend/onyx/connectors/canvas/connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:24]: Not sure what greptile is talking about but I looked thru the api docs and html_url doesn't exist. url is handled properly downstream as a slug.

File: backend/onyx/connectors/canvas/client.py (hunk 8 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py]: Seconds - added comment in code specifying

File: backend/onyx/connectors/canvas/client.py (hunk 5 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py]: added validation

File: backend/onyx/connectors/canvas/client.py (hunk 11 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py]: no, changed to >=400 and used onyx error codes

File: backend/onyx/connectors/canvas/connector.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+logger = setup_logger()
+
+
+class CanvasCourse(BaseModel):
+    id: int
+    name: str
+    course_code: str
+    created_at: str
+    workflow_state: str
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:14]: canvas api doesn't return html_url or url. link can be built with course id. I only use the course id to get further info from the course like assignments and pages to turn to docs.

File: backend/onyx/connectors/canvas/connector.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    due_at: str | None = None
+
+
+class CanvasAnnouncement(BaseModel):
+    id: int
+    title: str
+    message: str | None = None
+    html_url: str
+    posted_at: str | None = None
+    course_id: int
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:44]: canvas discussion topics api does not always guarantee created at, only posted at. use posted at for checkpoints, if no posted at, skipped by design

File: backend/onyx/connectors/canvas/connector.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:45]: no updated at in discussion topics api. only time-based thing is posted at

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
+"""Tests for Canvas connector — client (PR1)."""
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py:1]: only mediawiki has an init no other connectors do. not an issue

File: backend/onyx/connectors/canvas/connector.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    url: str
+    title: str
+    body: str | None = None
+    created_at: str
+    updated_at: str
+    course_id: int
+
+    @property
+    def id(self) -> int:
+        return self.page_id
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: valid, not used and already removed in pr2, removed this field.

File: backend/onyx/connectors/canvas/client.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+    r'<([^>]+)>;\s*rel="next"'
+)
+
+
+_STATUS_TO_ERROR_CODE: dict[int, OnyxErrorCode] = {
+    401: OnyxErrorCode.CREDENTIAL_EXPIRED,
+    403: OnyxErrorCode.INSUFFICIENT_PERMISSIONS,
+    404: OnyxErrorCode.BAD_GATEWAY,
+    429: OnyxErrorCode.RATE_LIMITED,
+}
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py:30]: no other connectors do this

File: backend/onyx/connectors/canvas/client.py [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+        bearer_token: str,
+        canvas_base_url: str,
+    ) -> None:
+        parsed_base = urlparse(canvas_base_url)
+        if not parsed_base.hostname:
+            raise ValueError("canvas_base_url must include a valid host")
+        if parsed_base.scheme != "https":
+            raise ValueError("canvas_base_url must use https")
+
+        self._bearer_token = bearer_token
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py:57]: Should we do any validation on the bearer_token that is passed in or is this happening upstream?

File: backend/onyx/connectors/canvas/client.py [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+        )
+
+        try:
+            response_json = response.json()
+        except ValueError as e:
+            if response.status_code < 300:
+                raise OnyxError(
+                    OnyxErrorCode.BAD_GATEWAY,
+                    detail=f"Invalid JSON in Canvas response: {e}",
+                )
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py:102]: This is confusing logic. Are we saying that we will try to get the json response but if that fails & is less than 300 status code we raise an error? What status codes would fall in here? the 2XX status codes are usually successful status codes.

File: backend/onyx/connectors/canvas/client.py [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+                        "Canvas pagination returned an unexpected host "
+                        f"({parsed_url.hostname}); expected {expected_host}"
+                    ),
+                )
+            if parsed_url.scheme != "https":
+                raise OnyxError(
+                    OnyxErrorCode.BAD_GATEWAY,
+                    detail=(
+                        "Canvas pagination link must use https, "
+                        f"got {parsed_url.scheme!r}"
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/client.py:169]: Odd syntax so make sure you understand how this log is happening

File: backend/onyx/connectors/canvas/client.py [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+
+    def _build_url(self, endpoint: str) -> str:
+        """Build a full Canvas API URL from an endpoint path.
+
+        Assumes endpoint is non-empty (e.g. ``"courses"``, ``"announcements"``).
+        Only called on a first request, endpoint must be set for first request.
+        Leading slashes are stripped to avoid double-slash in the result.
+        self.base_url is already normalized with no trailing slash.
+        """
+        return self.base_url + "/" + endpoint.lstrip("/")
```
[the remark this answers is not in the document]
Comment (justin-tahara) [reply backend/onyx/connectors/canvas/client.py]: @benwu408 can you properly confirm if this valid?

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+
+        assert result == OnyxErrorCode.BAD_GATEWAY
+
+    def test_503_maps_to_bad_gateway(self) -> None:
+        result = _error_code_for_status(503)
+
+        assert result == OnyxErrorCode.BAD_GATEWAY
+
+    def test_unknown_4xx_maps_to_bad_gateway(self) -> None:
+        result = _error_code_for_status(418)
```
Comment (justin-tahara) [inline backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: More of a personal preference but test names like these are a bit odd. I think being able to quickly see the status is good but the main thing is that you know what errors are being produced. 

Also for these tests, it feels redundant as these should exist in the OnyxErrorCode files sets of tests and not here.

This file should be capturing tests for your connector

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+    def test_returns_bearer_auth(self) -> None:
+        client = CanvasApiClient(
+            bearer_token=FAKE_TOKEN,
+            canvas_base_url=FAKE_BASE_URL,
+        )
+
+        result = client._build_headers()
+        expected = {"Authorization": f"Bearer {FAKE_TOKEN}"}
+
+        assert result == expected
```
Comment (justin-tahara) [inline backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: I might be wrong but since you don't do any validation on the bearer token I think you would get a weird output and that kind of stuff should ideally be tested here and you should make sure to test that edge case

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py [hunk 25]
```diff
@@ -0,0 +1,444 @@
...
+    @patch("onyx.connectors.canvas.client.rl_requests")
+    def test_success_returns_json_and_next_url(self, mock_requests: MagicMock) -> None:
+        next_link = f"<{FAKE_BASE_URL}/api/v1/courses?page=2>; " 'rel="next"'
+        mock_requests.get.return_value = _mock_response(
+            json_data=[{"id": 1}], link_header=next_link
+        )
+        client = CanvasApiClient(
+            bearer_token=FAKE_TOKEN,
+            canvas_base_url=FAKE_BASE_URL,
+        )
```
Comment (justin-tahara) [inline backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: A lot of these tests are creating the same client over and over again. maybe consider using a proper setUp() and tearDown() function for things that each test will need and it will honestly make this file much smaller since it removes the duplicated code

File: backend/onyx/connectors/canvas/client.py (hunk 19 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py:57]: load credentials is implemented in PR2 and it makes a test call

File: backend/onyx/connectors/canvas/client.py (hunk 20 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py:102]: yeah that part's in the except after response_json = response.json() which is where it tries to parse the response. so if it returns a successful code but can't parse we raise an error to say that the response came thru but was invalid json format

File: backend/onyx/connectors/canvas/client.py (hunk 21 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py:169]: 👍

Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py:169]: just checks host name and then if prefix is https and either of those aren't true it tells user whats wrong with the link

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py (hunk 23 above)
Comment (benwu408) [reply backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: ok so I made the function error_code_for_status that just maps an error code number to an actual onyx error code. but some of these were already tested in test_get so I just added two more to make sure all errors are tested in test_get and deleted this class.

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py (hunk 24 above)
Comment (benwu408) [reply backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: since we verify token in PR2 I have tests:

test_canvas_client_raises_without_credentials

test_load_credentials_invalid_token

on pr2 tests

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py (hunk 25 above)
Comment (benwu408) [reply backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: good point, implemented

File: backend/onyx/connectors/canvas/client.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/client.py]: oh found it heres where I can reply lol
